File: web/src/services/sandbox.js
/**
 * 沙盒运行服务
 * 负责与 /api/sandbox 相关接口交互
 */

import httpRequest from './request';

/**
 * 创建一次沙盒运行任务
 * @param {Object} payload
 * @returns {Promise<Object>}
 */
export async function createSandboxRun(payload) {
  const response = await httpRequest.post('/sandbox/runs', payload);
  return response?.task_id ? response : response?.data || response;
}

/**
 * 查询沙盒任务详情
 * @param {string} taskId
 * @returns {Promise<Object>}
 */
export async function fetchSandboxRun(taskId) {
  if (!taskId) {
    return null;
  }
  const response = await httpRequest.get(`/sandbox/runs/${taskId}`);
  return response?.task_id ? response : response?.data || response;
}

/**
 * 停止沙盒任务
 * @param {string} taskId
 * @returns {Promise<Object>}
 */
export async function stopSandboxRun(taskId) {
  const response = await httpRequest.post(`/sandbox/runs/${taskId}/stop`);
  return response?.task_id ? response : response?.data || response;
}

/**
 * 获取阶段日志内容
 * @param {string} taskId
 * @param {string} stage
 * @returns {Promise<Object>}
 */
export async function fetchSandboxStageLogs(taskId, stage) {
  const response = await httpRequest.get(`/sandbox/runs/${taskId}/logs/${stage}`);
  return response?.stage ? response : response?.data || response;
}

/**
 * 查询某个发布最近一次沙盒结果
 * @param {number|string} releaseId
 * @returns {Promise<Object|null>}
 */
export async function fetchLatestSandboxRun(releaseId) {
  if (!releaseId) {
    return null;
  }
  const response = await httpRequest.get(`/releases/${releaseId}/sandbox/latest`);
  return response?.task_id ? response : response?.data || response;
}

/**
 * 获取沙盒历史记录
 * @param {number|string} releaseId
 * @param {number} limit
 * @returns {Promise<Object>}
 */
export async function fetchSandboxHistory(releaseId, limit = 20) {
  if (!releaseId) {
    return { total: 0, items: [] };
  }
  const response = await httpRequest.get(`/releases/${releaseId}/sandbox/runs`, {
    params: { limit },
  });
  const data = response?.items ? response : response?.data || response;
  return data ?? { total: 0, items: [] };
}

File: src/api/sandbox.rs
use std::str::FromStr;

use actix_web::{HttpResponse, get, post, web};

use crate::error::AppError;
use crate::server::{
    CreateSandboxRunRequest, SandboxStage, SandboxState, create_sandbox_run_logic,
    get_latest_sandbox_run_logic, get_sandbox_run_logic, get_stage_logs_logic,
    list_sandbox_history_logic, stop_sandbox_run_logic,
};

/// 路径参数：task_id。
#[derive(serde::Deserialize)]
pub struct SandboxTaskPath {
    pub task_id: String,
}

/// 路径参数：task_id + 阶段。
#[derive(serde::Deserialize)]
pub struct SandboxLogPath {
    pub task_id: String,
    pub stage: String,
}

/// 路径参数：release id。
#[derive(serde::Deserialize)]
pub struct SandboxReleasePath {
    pub id: i32,
}

/// 历史记录查询参数。
#[derive(serde::Deserialize)]
pub struct SandboxHistoryQuery {
    pub limit: Option<u32>,
}

/// 默认的历史记录条数。
fn default_history_limit() -> u32 {
    20
}

/// 创建沙盒运行任务
#[post("/api/sandbox/runs")]
pub async fn create_sandbox_run(
    state: web::Data<SandboxState>,
    req: web::Json<CreateSandboxRunRequest>,
) -> Result<HttpResponse, AppError> {
    let resp = create_sandbox_run_logic(state.get_ref().clone(), req.into_inner()).await?;
    Ok(HttpResponse::Accepted().json(resp))
}

/// 查询沙盒任务详情
#[get("/api/sandbox/runs/{task_id}")]
pub async fn get_sandbox_run(
    state: web::Data<SandboxState>,
    path: web::Path<SandboxTaskPath>,
) -> Result<HttpResponse, AppError> {
    let run = get_sandbox_run_logic(state.get_ref().clone(), &path.task_id).await?;
    Ok(HttpResponse::Ok().json(run))
}

/// 停止沙盒任务
#[post("/api/sandbox/runs/{task_id}/stop")]
pub async fn stop_sandbox_run(
    state: web::Data<SandboxState>,
    path: web::Path<SandboxTaskPath>,
) -> Result<HttpResponse, AppError> {
    let run = stop_sandbox_run_logic(state.get_ref().clone(), &path.task_id).await?;
    Ok(HttpResponse::Ok().json(run))
}

/// 获取指定阶段日志
#[get("/api/sandbox/runs/{task_id}/logs/{stage}")]
pub async fn get_sandbox_stage_logs(
    state: web::Data<SandboxState>,
    path: web::Path<SandboxLogPath>,
) -> Result<HttpResponse, AppError> {
    let stage = SandboxStage::from_str(&path.stage)
        .map_err(|_| AppError::validation("无法解析阶段参数"))?;
    let resp = get_stage_logs_logic(state.get_ref().clone(), &path.task_id, stage).await?;
    Ok(HttpResponse::Ok().json(resp))
}

/// 获取 Release 最近一次沙盒结果
#[get("/api/releases/{id}/sandbox/latest")]
pub async fn get_latest_sandbox_run(
    state: web::Data<SandboxState>,
    path: web::Path<SandboxReleasePath>,
) -> Result<HttpResponse, AppError> {
    let resp = get_latest_sandbox_run_logic(state.get_ref().clone(), path.id).await?;
    Ok(HttpResponse::Ok().json(resp))
}

/// 获取指定发布的沙盒历史记录
#[get("/api/releases/{id}/sandbox/runs")]
pub async fn list_sandbox_history(
    state: web::Data<SandboxState>,
    path: web::Path<SandboxReleasePath>,
    query: web::Query<SandboxHistoryQuery>,
) -> Result<HttpResponse, AppError> {
    let limit = query
        .limit
        .unwrap_or_else(default_history_limit)
        .clamp(1, 100) as u64;
    let resp = list_sandbox_history_logic(state.get_ref().clone(), path.id, limit).await?;
    Ok(HttpResponse::Ok().json(resp))
}
